_source/_posts/2021-08-13-reactive-java.md

```diff
@@ -0,0 +1,791 @@
+Run with:
+```shell
+./mvnw test -Dtest=SecureRandomControllerTest
+```
+The test should pass, but **how can you make sure the REST call won't freeze up the service's event loop?** Use [BlockHound](https://github.com/reactor/BlockHound)!.,. Blockhound is a Java agent that detects blocking calls from non-blocking threads, has built-in integration with Project Reactor, and supports the JUnit platform.
+
+Add the Blockhound dependency to the `pom.xml`:
+
+```xml
+<dependency>
+    <groupId>io.projectreactor.tools</groupId>
+    <artifactId>blockhound-junit-platform</artifactId>
+    <version>1.0.6.RELEASE</version>
+    <scope>test</scope>
+</dependency>
+```
+
+It has been reported that when [`spring-security` is enabled, BlockHound does not detect](https://github.com/reactor/BlockHound/issues/173) blocking calls.
+
+Disable security for the test profile. Add the file `src/test/resources/application-test.yml` with the following content:
+
+```
+spring:
+  autoconfigure:
+    exclude:
+      - org.springframework.boot.autoconfigure.security.oauth2.client.reactive.ReactiveOAuth2ClientAutoConfiguration
+      - org.springframework.boot.autoconfigure.security.oauth2.resource.reactive.ReactiveOAuth2ResourceServerAutoConfiguration
+      - org.springframework.boot.autoconfigure.security.reactive.ReactiveSecurityAutoConfiguration
+```
+
+Rerun the test, and you should see the following error:
+
+```txt
+reactor.blockhound.BlockingOperationError: Blocking call! java.io.FileInputStream#readBytes
+	at java.base/java.io.FileInputStream.readBytes(FileInputStream.java) ~[na:na]
+	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException:
+Error has been observed at the following site(s):
+	|_ checkpoint ⇢ HTTP GET "/random" [ExceptionHandlingWebHandler]
+Stack trace:
+		at java.base/java.io.FileInputStream.readBytes(FileInputStream.java) ~[na:na]
+		at java.base/java.io.FileInputStream.read(FileInputStream.java:279) ~[na:na]
+		at java.base/java.io.FilterInputStream.read(FilterInputStream.java:133) ~[na:na]
+		at java.base/sun.security.provider.NativePRNG$RandomIO.readFully(NativePRNG.java:424) ~[na:na]
+		at java.base/sun.security.provider.NativePRNG$RandomIO.ensureBufferValid(NativePRNG.java:526) ~[na:na]
+		at java.base/sun.security.provider.NativePRNG$RandomIO.implNextBytes(NativePRNG.java:545) ~[na:na]
+		at java.base/sun.security.provider.NativePRNG.engineNextBytes(NativePRNG.java:220) ~[na:na]
+		at java.base/java.security.SecureRandom.nextBytes(SecureRandom.java:741) ~[na:na]
+		at java.base/java.security.SecureRandom.next(SecureRandom.java:798) ~[na:na]
+```
+
+As the log above shows, `SecureRandom.next` produces a call to `FileInputStream#readBytes`, which is blocking.
+
+The `SecureRandomService` returns a `Mono` publisher, but it is an **Impostor Reactive Service**!
+
+```java
+return Mono.just(secureRandom.nextInt());
+```
+
+The implementation above has logic upfront, first the calculation `secureRandom.nextInt` and then the assembly.
```

Comment:
```suggestion
The implementation above has logic up front, first the calculation `secureRandom.nextInt` and then the assembly.
```
because it's an adverb here, not an adj.